### PR TITLE
Add methods to form thermometer codes

### PR DESCRIPTION
**Pull request by denkle**

### CODE REVIEW

**Review by mikeheddes — COMMENTED** [review]
Thank you for the PR, it's looking good! I just have a couple comments about some details.

File: torchhd/functional.py [hunk 1]
```diff
@@ -325,6 +326,85 @@ def level_hv(
...
+def thermometer_hv(
+    dimensions: int,
+    model: Type[VSA_Model] = MAP,
+    *,
+    requires_grad=False,
+    **kwargs,
+) -> VSA_Model:
+    """Creates a thermometer code for given dimensionality.
+
+    Implements similarity-preserving hypervectors as described in "Sparse Binary Distributed Encoding of Scalars" <https://doi.org/10.1615/J Automat Inf Scien.v37.i6.20>.
```
[the remark this answers is not in the document]
Comment (denkle) [reply torchhd/functional.py:338]: Thanks. I have not thought about the documentation in the first place. Fixed it now by updating the corresponding .rst files. 
There was indeed an error in the link formatting. Fixed it as well so it works in the generated .html

File: torchhd/embeddings.py [hunk 2]
```diff
@@ -168,6 +169,64 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:
...
+        )
+
+        self._fill_padding_idx_with_zero()
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        indices = functional.value_to_index(
+            input, self.low_value, self.high_value, self.num_embeddings
+        ).clamp(0, self.num_embeddings - 1)
+
+        return super(Thermometer, self).forward(indices).as_subclass(MAP)
```
Comment (denkle) [inline torchhd/embeddings.py:229]: @mikeheddes a question here. I followed the structure as in Level(). It, however, looks weird that MAP is used at the end when doing .as_subclass(). It seems that we rather need to specify a "model" parameter at the input and use the corresponding model with return.

File: torchhd/functional.py [hunk 3]
```diff
@@ -325,6 +326,85 @@ def level_hv(
...
+
+
+    hv = torch.empty(
+        num_vectors,
+        dimensions,
+        dtype=rand_hv.dtype,
+        device=rand_hv.device,
+    )
+
+    for i in range(num_vectors):
```
[the remark this answers is not in the document]
Comment (denkle) [reply torchhd/functional.py]: Great point! I agree it is no-brainer that the usage of indexing will make the implementation extremely fast. So the code was modified accordingly

File: torchhd/functional.py [hunk 4]
```diff
@@ -325,6 +326,85 @@ def level_hv(
...
     return hv.as_subclass(model)
 
+def thermometer_hv(
+    dimensions: int,
+    model: Type[VSA_Model] = MAP,
+    *,
+    requires_grad=False,
+    **kwargs,
+) -> VSA_Model:
+    """Creates a thermometer code for given dimensionality.
```
[the remark this answers is not in the document]
Comment (denkle) [reply torchhd/functional.py:336]: We need to save this point somewhere! I think the current implementation with the use of torch.tril_indices is pretty compact but as we expand the number of supported HD/VSA models we might want to switch to this idea of interpolation.

File: torchhd/embeddings.py (hunk 2 above)
Comment (mikeheddes) [reply torchhd/embeddings.py:229]: Yes, you are right. It is currently this way to be compatible with the way we implemented it before v4. But we should allow the user to specify the model of the various embedding classes. I will make a new issue for this.

**Comment by mikeheddes** [comment]
I'm wondering whether it would be better to keep the same API as for the other hypervector creation functions where you can specify the number of vectors and their dimensions. Because for 10000 dimension this will create a very large matrix while you might only need to use 6 hypervectors or something.

**Comment by denkle** [comment]
> I'm wondering whether it would be better to keep the same API as for the other hypervector creation functions where you can specify the number of vectors and their dimensions. Because for 10000 dimension this will create a very large matrix while you might only need to use 6 hypervectors or something.
Will work now on revising the code to address this.

**Review by mikeheddes — APPROVED** [review]
@denkle thank you for revising the code and congrats with merging your first PR! 

I removed the VSA model argument from the embedding class so that all the embeddings have a consistent API. I think it's better if we change all the embedding classes to accept a VSA model argument in a separate PR.

**Comment by denkle** [comment]
Thanks, @mikeheddes! Agree with the point on separate PR for changing embeddings' API for all classes. Just let me know if you would like to start it or whether I should give the first try. It might at least be worth checking if the code for Thermometer was reasonable.

**Comment by mikeheddes** [comment]
@denkle you can start working on that feature if you want. What you had before was in the right direction I think, it's a similar API to the functional ones. The only thing that was still missing is making sure that the dtype of the created tensors matches (or is compatible with) the provided VSA_Model.

**Pull request merged**